clang/lib/Sema/SemaInit.cpp

```diff
@@ -2363,8 +2356,11 @@ void InitListChecker::CheckStructUnionTypes(
   }
 
   // Emit warnings for missing struct field initializers.
-  if (!VerifyOnly && InitializedSomething && CheckForMissingFields &&
-      !RD->isUnion()) {
+  // This check is disabled for designated initializers in C.
+  // This matches gcc behaviour.
+  if (!VerifyOnly && InitializedSomething && !RD->isUnion() &&
+      !IList->isIdiomaticZeroInitializer(SemaRef.getLangOpts()) &&
+      !(HasDesignatedInit && !SemaRef.getLangOpts().CPlusPlus)) {
```

Comment:
```suggestion
      (!HasDesignatedInit || SemaRef.getLangOpts().CPlusPlus)) {
```

Reply:
Minor preference for extracting the condition out into well-named bool variables, this is getting to be a little bit of a mess.

Reply:
Unexpanded condition corresponds to the comment I left a few lines above:
> This check is disabled for designated initializers in C.

For me personally, it looks a bit less readable in expanded form ("check unused fields if there are no designated initializers or if language is C++")

Reply:
@erichkeane, I did something like this before 83864610d6b6470a0f229c516821d8344d994562, but this adds another indentation level.

It's possible to calculate `DisableCheck` before the `if` statement, but it'll probably be less optimal than the current solution - if `VerifyOnly` is true, `DisableCheck` is not needed.

Reply:
> I am asking for the list of '!'s to be shorter/grouped into a couple of 'bools' with descriptive names

Like this? Other checks seem unrelated to each other.
```diff
+ bool IsCDesignatedInitializer = HasDesignatedInit && !SemaRef.getLangOpts().CPlusPlus;
  if (!VerifyOnly && InitializedSomething && !RD->isUnion() &&
      !IList->isIdiomaticZeroInitializer(SemaRef.getLangOpts()) &&
-     !(HasDesignatedInit && !SemaRef.getLangOpts().CPlusPlus)) {
+     !IsCDesignatedInitializer) {
```

Reply:
That one definitely helps... I was hoping they grouped better.  But I guess if that is what we can get.  Basically my issue is how difficult it was to deduce what I was looking at with all hte conditions.  Perhaps Aaron has a suggestion.